llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -394,34 +398,40 @@ class PropagateIncomingLatencies : public ScheduleDAGMutation {
       const MachineBasicBlock &MBB = *MI.getParent();
       const MachineRegisterInfo &MRI = DAG->MRI;
       auto HasExternalAndLocalSources = [&MBB, &MRI](const MachineInstr &MI) {
-        return MI.isRegSequence() && MRI.isSSA() && MI.getNumOperands() > 3 &&
-               count_if(MI.uses(), [&MBB, &MRI](const MachineOperand &MO) {
-                 return MO.isReg() && MO.getReg().isVirtual() &&
-                        MRI.getVRegDef(MO.getReg())->getParent() != &MBB;
-               }) == 1;
+        if (!MI.isRegSequence() || !MRI.isSSA())
+          return false;
+        const auto NumExternal =
+            count_if(MI.uses(), [&MBB, &MRI](const MachineOperand &MO) {
+              return MO.isReg() && MO.getReg().isVirtual() &&
+                     MRI.getVRegDef(MO.getReg())->getParent() != &MBB;
+            });
+        const auto NumInternal = MI.getNumOperands() - 1 - (2 * NumExternal);
+        return NumExternal == 1 && NumInternal >= 1;
```

Comment:
Here, do we consider a subregister index of an internal value as accounting for `NumInternal` ? Should we divide `NumInternal` by 2?

Reply:
I'm not completely clear on what would be the best heuristic tbh. As mentioned here https://github.com/Xilinx/llvm-aie/pull/220#discussion_r1818591503 I have experimented a bit, but it is challenging to find something that consistently yields good results. The current code works pretty well (see results in PR description), and I'm a bit afraid to specialise the heuristic too much for our current benchmarks if we keep tweaking it. I would suggest leaving that basic heuristic intact, and tweak it in follow-up work as we see fit. What do you think?